### PR TITLE
Speed up test runs by leveraging in-memory databases within reason

### PR DESCRIPTION
**Pull request by AmineKhaldi**

### CODE REVIEW

File: chia/simulator/setup_services.py [hunk 1]
```diff
@@ -119,17 +119,18 @@ async def setup_full_node(
...
-        # TODO: remove (maybe) when fixed https://github.com/python/cpython/issues/97641
-        gc.collect()
-        db_path.unlink()
-
-        if db_version > 1:
-            with sqlite3.connect(db_path) as connection:
-                connection.execute("CREATE TABLE database_version(version int)")
-                connection.execute("INSERT INTO database_version VALUES (?)", (db_version,))
-                connection.commit()
+    if reuse_db:
```
Comment (arvidn) [inline chia/simulator/setup_services.py:122]: does `reuse_db` mean that we want to be able to close the connection, and then open up a new connection?
if it just means we want multiple connections to it, you can do that with an in-memory DB as well, you just need to add `&cache=shared` to the URI

File: tests/util/db_connection.py [hunk 2]
```diff
@@ -1,19 +1,32 @@
...
-        db_path = Path(directory).joinpath("db.sqlite")
-        _db_wrapper = await DBWrapper2.create(database=db_path, reader_count=4, db_version=db_version)
-        try:
-            yield _db_wrapper
-        finally:
-            await _db_wrapper.close()
+async def DBConnection(db_version: int, in_memory_db: bool = True) -> AsyncIterator[DBWrapper2]:
+    if in_memory_db:
+        directory = None
+        db_path: Union[str, Path] = gen_in_memory_db_uri("db_cnx_")
```
Comment (arvidn) [inline tests/util/db_connection.py:22]: if multiple tests open a DB connection to the same URI, won't they access the same underlying DB, and possibly interfere with each other?

File: tests/util/db_connection.py [hunk 3]
```diff
@@ -1,19 +1,32 @@
...
+    else:
+        directory = tempfile.TemporaryDirectory()
```
Comment (arvidn) [inline tests/util/db_connection.py:24]: I think an important benefit of using the context manager here is that the temporary directory *will* be cleaned up. in your version, if `_db_wrapper.close()` throws, it won't be

File: tests/util/db_connection.py [hunk 4]
```diff
@@ -1,19 +1,32 @@
...
 from pathlib import Path
-from typing import AsyncIterator
+from typing import AsyncIterator, Union
 
 from chia.util.db_wrapper import DBWrapper2
 
 
+def gen_in_memory_db_uri(prefix: str) -> str:
+    random_string = "".join(random.choice(string.ascii_letters) for _ in range(8))
+    return f"file:{prefix}{random_string}?mode=memory&cache=shared"
```
Comment (arvidn) [inline tests/util/db_connection.py:15]: I see you added `cache=shared` here, to make sure multiple DB connections will share the same underlying database. Is that the intention?

File: chia/simulator/setup_services.py [hunk 5]
```diff
@@ -119,17 +119,18 @@ async def setup_full_node(
...
+        db_path: Union[str, Path] = local_bt.root_path / f"{db_name}"
+        uri = False
+    else:
+        db_path = f"file:db_{db_name}?mode=memory"
```
Comment (arvidn) [inline chia/simulator/setup_services.py:126]: is there a reason to not include `&cache=shared` here?

**Comment by arvidn** [comment]
How much faster are the tests with this change?

File: tests/util/db_connection.py [hunk 6]
```diff
@@ -1,19 +1,32 @@
...
 from contextlib import asynccontextmanager
 from pathlib import Path
-from typing import AsyncIterator
+from typing import AsyncIterator, Union
 
 from chia.util.db_wrapper import DBWrapper2
 
 
+def gen_in_memory_db_uri(prefix: str) -> str:
+    random_string = "".join(random.choice(string.ascii_letters) for _ in range(8))
```
Comment (arvidn) [inline tests/util/db_connection.py:14]: given that you use a random string here, is there really a need to pass in the prefix?
this won't show up on a filesystem anyway, I don't think a user or developer will really encounter this string anyway

**Comment by AmineKhaldi** [comment]
> How much faster are the tests with this change?

I used `tests/wallet/test_wallet_trade_store.py::test_large_trade_record_query` locally as an example:

main: 1 passed in 141.55s
this: 1 passed in 124.67s

**Review by arvidn — APPROVED** [review]

**Comment by trepca** [comment]
@arvidn @AmineKhaldi  This introduces a dependency in `setup_simulators_and_wallets` to `tests.util`. Problem is that `tests.util` is not included in distro package, meaning my tests that are using simulators are not working anymore.

**Comment by AmineKhaldi** [comment]
> @arvidn @AmineKhaldi This introduces a dependency in `setup_simulators_and_wallets` to `tests.util`. Problem is that `tests.util` is not included in distro package, meaning my tests that are using simulators are not working anymore.

Yep, this PR got merged before adding this change to it, so I sent an addendum PR #16752 shortly after.
Good to see you seb!

**Pull request merged**